pkg/syncer/storagepool/diskDecommissionController.go

```diff
@@ -0,0 +1,277 @@
+				break
+			}
+			pvcToMigrate = append(pvcToMigrate, pvc)
+		}
+
+		_, unsuccessfulMigrations := w.migrationCntlr.MigrateVolumes(ctx, pvcToMigrate, true)
+		if len(unsuccessfulMigrations) != 0 {
+			migrationFailed = true
+		}
+	}
+}
+
+func initDiskDecommController(ctx context.Context, migrationCntlr *migrationController) (*DiskDecommController, error) {
+	log := logger.GetLogger(ctx)
+	log.Infof("Starting disk decommission controller")
+	k8sDynamicClient, spResource, err := getSPClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	w := &DiskDecommController{}
+	w.k8sDynamicClient = k8sDynamicClient
+	w.migrationCntlr = migrationCntlr
+	w.spResource = spResource
+	w.pvResource = &schema.GroupVersionResource{Group: "", Version: "v1", Resource: "persistentvolumes"}
+	w.pvcResource = &schema.GroupVersionResource{Group: "", Version: "v1", Resource: "persistentvolumeclaims"}
+	w.diskDecommMode = make(map[string]string)
+	w.execSemaphore = semaphore.NewWeighted(1)
+
+	// get all the pvc resource for which targetSPAnnotationKey annotations exists.
+	// This will give us list of all pending migrations.
+	pvcList, err := w.k8sDynamicClient.Resource(*w.pvcResource).Namespace(v1.NamespaceAll).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return w, err
+	}
+
+	pvcToMigrate := make([]*unstructured.Unstructured, 0)
+	for _, pvc := range pvcList.Items {
+		targetSPName, _, _ := unstructured.NestedString(pvc.Object, "metadata", "annotations", targetSPAnnotationKey)
+		if targetSPName != "" {
+			pvcToMigrate = append(pvcToMigrate, &pvc)
+		}
+	}
+	w.migrationCntlr.MigrateVolumes(ctx, pvcToMigrate, false)
+
+	// start StoragePool watch to look for events putting SP under disk decommission
+	err = w.renewStoragePoolWatch(ctx)
+	if err != nil {
+		return w, err
+	}
+	go w.watchStoragePool(ctx)
+
+	// get all sp resource
+	spList, err := w.k8sDynamicClient.Resource(*w.spResource).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return w, err
+	}
+	// make progress on StoragePool which are under disk decommission
+	for _, sp := range spList.Items {
+		spName := sp.GetName()
+		if w.shouldEnterDiskDecommission(ctx, &sp) && spName != "" {
+			maintenanceMode := w.diskDecommMode[spName]
+			go w.DecommissionDisk(ctx, spName, maintenanceMode)
+		}
+	}
+	return w, nil
+}
+
+// As our watch can and will expire, we need a helper to renew it. Note that after we re-new it,
+// we will get a bunch of already processed events.
+func (w *DiskDecommController) renewStoragePoolWatch(ctx context.Context) error {
+	log := logger.GetLogger(ctx)
+	spClient, spResource, err := getSPClient(ctx)
+	if err != nil {
+		return err
+	}
+	// This means every 24h our watch may expire and require to be re-created.
+	// When that happens, we may need to do a full remediation, hence we change
+	// from 30m (default) to 24h.
+	timeout := int64(60 * 60 * 24) // 24h
+	w.spWatch, err = spClient.Resource(*spResource).Watch(ctx, metav1.ListOptions{
+		TimeoutSeconds: &timeout,
```

Comment:
Not sure but do we have Timeoutinhours to simplify this?

Reply:
No, metav1.ListOptions only has TimeoutSeconds field to specify timeout.